### PR TITLE
Implement manual retention sweep

### PR DESCRIPTION
**Pull request by rbev**
This pull request introduces a manual data retention sweep API for ServiceControl, enabling users to trigger and monitor retention sweeps via new endpoints. It includes new API contracts, permission management, and enhancements to the EFCore persistence layer to support manual and scheduled sweeps, including proper status reporting and concurrency control. Additionally, it ensures that unsupported operations on RavenDB are handled gracefully.

**Manual Retention Sweep API and EFCore Support:**

* Introduced new API contracts: `RetentionSweepRequest`, `RetentionSweepResponse`, and `RetentionSweepStatus` to enable manual triggering and status polling for data retention sweeps (`src/ServiceControl.Api/Contracts/RetentionSweepRequest.cs`, `RetentionSweepResponse.cs`, `RetentionSweepStatus.cs`). [[1]](diffhunk://#diff-632a1156888d131c782e10a96c28185eff6db8391e6a37be192ac81dd2e2ac04R1-R23) [[2]](diffhunk://#diff-f1dd8128580a129ac3deb424ed28f68437a39a624baa4c67e869dd5d2152763dR1-R22) [[3]](diffhunk://#diff-98e4b5f70e09ea8aea259ca457bfa21cf02563950f6e76966248e6ec5aae9787R1-R25)
* Added the `IRetentionApi` interface defining methods for starting a sweep and retrieving sweep status, with support for indicating when the operation is not supported (e.g., on RavenDB) (`src/ServiceControl.Api/IRetentionApi.cs`).
* Enhanced the EFCore `RetentionSweeper` to support manual sweeps with caller-supplied cutoffs, concurrency control to prevent overlapping sweeps, and detailed status reporting for the new API (`src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs`, `BasePersistence.cs`). [[1]](diffhunk://#diff-da84788333b253b21e0cc71e12446fad927888f74abb8672d9b563ada96efdd4R8) [[2]](diffhunk://#diff-da84788333b253b21e0cc71e12446fad927888f74abb8672d9b563ada96efdd4R17-R49) [[3]](diffhunk://#diff-da84788333b253b21e0cc71e12446fad927888f74abb8672d9b563ada96efdd4L43-R64) [[4]](diffhunk://#diff-da84788333b253b21e0cc71e12446fad927888f74abb8672d9b563ada96efdd4L61-R152) [[5]](diffhunk://#diff-da84788333b253b21e0cc71e12446fad927888f74abb8672d9b563ada96efdd4L108-R196) [[6]](diffhunk://#diff-da84788333b253b21e0cc71e12446fad927888f74abb8672d9b563ada96efdd4L138-R225) [[7]](diffhunk://#diff-cf186271d78dd3ad85666cee22bbdeb18b962bcd1750b3a11e098b74239accc1L43-R48)

**API Permissions and Roles:**

* Added a new permission constant `ErrorRetentionSweep` and included it in the admin role permissions to control access to the manual retention sweep API (`src/ServiceControl.Infrastructure/Auth/Permissions.cs`, `RolePermissions.cs`). [[1]](diffhunk://#diff-6d887e70b6cdaa98ba1a955963daf89bdfd5bde1f5b883c6a37cd2cfd1dba2c8R61-R63) [[2]](diffhunk://#diff-bccc85a1c347b91c2d1651f3b3ebad4a93bbe8d2cf9ade4a8edc0240d1f18effR66)

### CODE REVIEW

File: src/ServiceControl/Retention/Api/RetentionController.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    [Authorize(Policy = Permissions.ErrorRetentionSweep)]
+    [Route("retention/sweep")]
+    [HttpPost]
+    public async Task<IActionResult> Sweep([FromBody] RetentionSweepRequest request, CancellationToken cancellationToken = default)
+    {
+        var response = await retentionApi.SweepAsync(request ?? new RetentionSweepRequest(), cancellationToken);
+
+        return response.Status switch
+        {
+            "started" => Accepted(response),
```
Comment (rbev) [inline src/ServiceControl/Retention/Api/RetentionController.cs]: these statuses should probably be an enum

File: src/ServiceControl/Retention/Api/RetentionController.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+using Microsoft.AspNetCore.Mvc;
+using ServiceControl.Api;
+using ServiceControl.Api.Contracts;
+
+// Manual retention-sweep endpoint. Lives only on the primary error instance (the sweeper is only
+// registered there). On a RavenDB-backed instance IRetentionSweeper is not registered, so the
+// IRetentionApi implementation returns a "not-supported" status that this controller maps to 501.
+[ApiController]
+[Route("api")]
+public class RetentionController(IRetentionApi retentionApi) : ControllerBase
```
Comment (rbev) [inline src/ServiceControl/Retention/Api/RetentionController.cs]: Should this controller be more generic to allow for other sysadmin actions?

Perhaps `api/maintenence/retention/purge`

File: src/ServiceControl/Infrastructure/Api/RetentionApi.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        {
+            return Task.FromResult(NotSupported());
+        }
+
+        // Maintenance mode refuses mutating operations; a sweep while the DB is being maintained
+        // would contend with the maintenance work.
+        if (settings.PersisterSpecificSettings?.MaintenanceMode == true)
+        {
+            return Task.FromResult(new RetentionSweepResponse { Status = "maintenance", Reason = "The instance is in maintenance mode." });
+        }
```
Comment (johnsimons) [inline src/ServiceControl/Infrastructure/Api/RetentionApi.cs]: this may need to change given @warwickschroeder latest PRs

File: src/ServiceControl.Api/IRetentionApi.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    /// Starts a manual retention sweep with caller-supplied cutoffs. The delete work runs in
+    /// the background on a host-lifetime token; this method returns as soon as the run is
+    /// accepted (or refused because one is already running / unsupported).
+    /// </summary>
+    Task<RetentionSweepResponse> SweepAsync(RetentionSweepRequest request, CancellationToken cancellationToken = default);
+
+    /// <summary>
+    /// Returns a point-in-time snapshot of sweep execution state for polling.
+    /// </summary>
+    Task<RetentionSweepStatus> GetStatusAsync(CancellationToken cancellationToken = default);
```
Comment (johnsimons) [inline src/ServiceControl.Api/IRetentionApi.cs]: We don't tend to use the postfix "Async", the return gives it away

**Review by johnsimons — APPROVED** [review]

File: src/ServiceControl.Api/IRetentionApi.cs (hunk 4 above)
Comment (rbev) [reply src/ServiceControl.Api/IRetentionApi.cs]: Yeah, its a hard to shake habit after so many years of doing it.

File: src/ServiceControl/Retention/Api/RetentionController.cs (hunk 1 above)
Comment (rbev) [reply src/ServiceControl/Retention/Api/RetentionController.cs]: These have been moved to constants

File: src/ServiceControl/Retention/Api/RetentionController.cs (hunk 2 above)
Comment (rbev) [reply src/ServiceControl/Retention/Api/RetentionController.cs]: renamed

File: src/ServiceControl/Infrastructure/Api/RetentionApi.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        // a RavenDB-backed instance would throw at resolve time. Absent => 501 Not Implemented.
+        var sweeper = serviceProvider.GetService<IRetentionSweeper>();
+        if (sweeper is null)
+        {
+            return Task.FromResult(NotSupported());
+        }
+
+        // Maintenance mode refuses mutating operations; a sweep while the DB is being maintained
+        // would contend with the maintenance work.
+        if (settings.PersisterSpecificSettings?.MaintenanceMode == true)
```
Comment (warwickschroeder) [inline src/ServiceControl/Infrastructure/Api/RetentionApi.cs]: Can probably remove this check completely. SQL wont have any maintenance mode, and raven will fail earlier if this endpoint is called anyway.

File: src/ServiceControl/Infrastructure/Api/RetentionApi.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            return Task.FromResult(new RetentionPurgeResponse { Status = StatusMaintenance, Reason = "The instance is in maintenance mode." });
+        }
+
+        request ??= new RetentionPurgeRequest();
+
+        // Cutoffs must be UTC and in the past. A future cutoff would delete nothing and is almost
+        // certainly a caller mistake, so it is rejected rather than clamped.
```
Comment (warwickschroeder) [inline src/ServiceControl/Infrastructure/Api/RetentionApi.cs]: Also might be worth guarding this with a test.

```suggestion
        // Cutoffs must be UTC and in the past. A future cutoff matches every row and is almost
        // certainly a caller mistake, so it is rejected rather than clamped.
```

File: src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs [hunk 7]
```diff
@@ -13,19 +14,37 @@ namespace ServiceControl.Persistence.EFCore.Infrastructure;
...
+
+    // Status snapshot for GET /api/maintenance/retention/purge/status polling. Volatile reads/writes are
+    // sufficient here: the fields are written under sweepLock (or once at start) and read
+    // lock-free for status reporting, which only needs an eventually-consistent snapshot.
+    volatile bool isRunning;
+    DateTime? lastStartedAt;
+    DateTime? lastFinishedAt;
+    DateTime? lastErrorCutoff;
+    DateTime? lastEventsCutoff;
+    string? lastError;
```
Comment (warwickschroeder) [inline src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs]: Nothing seems to set this to a value other than null.

File: src/ServiceControl/Retention/Api/SystemMaintenanceController.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+public class SystemMaintenanceController(IRetentionApi retentionApi) : ControllerBase
+{
+    // Starts a full retention purge with caller-supplied cutoffs. The delete work runs in the
+    // background on a host-lifetime token; this returns as soon as the run is accepted (202),
+    // already running (409), in maintenance mode (503), unsupported by the persister (501), or
+    // the cutoff was invalid (400).
+    [Authorize(Policy = Permissions.ErrorRetentionPurge)]
+    [Route("retention/purge")]
+    [HttpPost]
+    public async Task<IActionResult> Purge([FromBody] RetentionPurgeRequest request, CancellationToken cancellationToken = default)
```
Comment (warwickschroeder) [inline src/ServiceControl/Retention/Api/SystemMaintenanceController.cs:26]: This is a pretty destructive API. We should audit who calls this with `ICurrentUserAccessor` and `IMessageActionAuditLog` at the very least. Another option is to only allow this API to be called when authentication and authorisation is enabled.

File: src/ServiceControl/Infrastructure/Api/RetentionApi.cs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        if (supplied is null)
+        {
+            validated = null;
+            error = null;
+            return true;
+        }
+
+        var value = supplied.Value;
+
+        if (value.Kind != DateTimeKind.Utc)
```
Comment (warwickschroeder) [inline src/ServiceControl/Infrastructure/Api/RetentionApi.cs]: `2026-01-01T00:00:00+00:00` and `2026-01-01T00:00:00` would be rejected. Can we normalise with `ToUniversalTime()` and only reject `Unspecified`?

File: src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs (hunk 7 above)
Comment (warwickschroeder) [inline src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs]: These seem to be shared between the manual purge and the hourly sweep. Is this going to cause a problem? Someone polling their manual, might get wrong values due to the hourly running? Or should the hourly be disabled while a manual is running, and vice versa?

File: src/ServiceControl/Retention/Api/SystemMaintenanceController.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        var response = await retentionApi.Sweep(request ?? new RetentionPurgeRequest(), cancellationToken);
+
+        return response.Status switch
+        {
+            RetentionApi.StatusStarted => Accepted(response),
+            RetentionApi.StatusAlreadyRunning => Conflict(response),
+            RetentionApi.StatusMaintenance => StatusCode(503, response),
+            RetentionApi.StatusNotSupported => StatusCode(501, response),
+            RetentionApi.StatusInvalidCutoff => BadRequest(response),
+            _ => Ok(response)
```
Comment (warwickschroeder) [inline src/ServiceControl/Retention/Api/SystemMaintenanceController.cs]: Any unrecognised status should be a success?

File: src/ServiceControl/Infrastructure/Api/RetentionApi.cs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+// same controller/route is mapped on every persister: EFCore registers it and gets 202/409/200;
+// RavenDB registers nothing (its retention is the server-side @expires bundle) and gets 501.
+class RetentionApi(IServiceProvider serviceProvider, Settings settings) : IRetentionApi
+{
+    public const string NotSupportedReason = "The currently configured storage has no retention sweeper.";
+    public const string StatusMaintenance = "maintenance";
+    public const string StatusStarted = "started";
+    public const string StatusNotSupported = "not-supported";
+    public const string StatusAlreadyRunning = "already-running";
+    public const string StatusInvalidCutoff = "invalid-cutoff";
```
Comment (warwickschroeder) [inline src/ServiceControl/Infrastructure/Api/RetentionApi.cs]: Do these public const belong here? Maybe as an enum in ServiceControl.Api/Contracts?

Comment (rbev) [reply src/ServiceControl/Infrastructure/Api/RetentionApi.cs]: Actually that's a good point, I was originally looking for a spot for the enum that could live in both API and persistence and couldnt' find one but no reason there couldn't be one here

File: src/ServiceControl/Infrastructure/Api/RetentionApi.cs (hunk 6 above)
Comment (rbev) [reply src/ServiceControl/Infrastructure/Api/RetentionApi.cs]: There is test coverage for these boundary cases: https://github.com/Particular/ServiceControl/pull/5845/changes#diff-8aa25e5f80a0269cc65655a73f045359332dcf478065da2e257d3ee1406feb6fR43

File: src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs (hunk 7 above)
Comment (rbev) [reply src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs]: They are, but the two have slightly different paths - 

* auto sweep blocks if another (manual) one is running
* manual run takes the lock and immediately and kicks off a background task to do the work

Both paths take the semaphore before mutating the variables and finish editing them before releasing it, however the manual run sets the start values before kicking off the background work.

File: src/ServiceControl/Infrastructure/Api/RetentionApi.cs (hunk 11 above)
Comment (rbev) [reply src/ServiceControl/Infrastructure/Api/RetentionApi.cs]: Done

File: src/ServiceControl/Infrastructure/Api/RetentionApi.cs (hunk 9 above)
Comment (rbev) [reply src/ServiceControl/Infrastructure/Api/RetentionApi.cs]: Fixed

File: src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs (hunk 7 above)
Comment (rbev) [reply src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs]: removed

File: src/ServiceControl/Retention/Api/SystemMaintenanceController.cs (hunk 8 above)
Comment (rbev) [reply src/ServiceControl/Retention/Api/SystemMaintenanceController.cs:26]: I added auditing

File: src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs [hunk 12]
```diff
@@ -58,13 +76,75 @@ protected override async Task ExecuteAsync(CancellationToken cancellationToken =
...
+
+        // Lock acquired on this thread. The background task owns it from here and releases it when
+        // the sweep body completes (SemaphoreSlim is not thread-affine, so releasing from the
+        // background thread is safe). isRunning is set now so a concurrent manual call sees it.
+        isRunning = true;
+        lastStartedAt = timeProvider.GetUtcNow().UtcDateTime;
+        lastErrorCutoff = errorCutoff;
+        lastEventsCutoff = eventsCutoff;
+
+        _ = SweepWithoutAcquiringLock();
```
Comment (johnsimons) [inline src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs:101]: Do we need to observe exceptions here?

File: src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs [hunk 13]
```diff
@@ -58,13 +76,75 @@ protected override async Task ExecuteAsync(CancellationToken cancellationToken =
...
+
+        return new ManualSweepAttempt(RetentionSweepStatus.Started, lastStartedAt, errorCutoff, eventsCutoff);
+
+        async Task SweepWithoutAcquiringLock()
+        {
+            try
+            {
+                // if the caller doesn't hand over a real cancellation token then use the application lifetime.
+                await SweepBody(errorCutoff, eventsCutoff, false, cancellationToken.CanBeCanceled ? cancellationToken : hostApplicationLifetime.ApplicationStopping);
```
Comment (johnsimons) [inline src/ServiceControl.Persistence.EFCore/Infrastructure/RetentionSweeper.cs]: Should we combine the tokens

**Review by johnsimons — APPROVED** [review]
just a couple of comments to consider

**Review by warwickschroeder — APPROVED** [review]
